### ISSUE TITLE
Include( x => x.Child.OrderBy ) does not work in Postgresql

### ISSUE THREAD

**Issue by jxwaters** [dotnet/efcore#29925]
I expect the following query to sort the child document collection by Created descending

`.Include(x => x.Documents.OrderByDescending( d => d.Created));`

but no order by is added to the generated SQL

```
SELECT t.start_date, t.end_date, t.content, t.audience, t.status, t.primary_button_name, t.primary_button_action, t.secondary_button_name, t.secondary_button_action, t.created, t.created_by, t.last_updated, t.last_updated_by, t.id, t0."Id", t0."DocumentType", t0."FileName", t0."MediaUrl", t0."FileSize", t.include_in_enterprise, t.header, t.sub_header
FROM (
    SELECT a.start_date, a.end_date, a.content, a.audience, a.status, a.primary_button_name, a.primary_button_action, a.secondary_button_name, a.secondary_button_action, a.created, a.created_by, a.last_updated, a.last_updated_by, a.include_in_enterprise, a.id, a.header, a.sub_header
    FROM master.announcements AS a
    WHERE a.is_active AND (a.id = @__id_0)
    LIMIT 2
) AS t
LEFT JOIN (
    SELECT a0.id AS "Id", a0.document_type AS "DocumentType", a0.file_name AS "FileName", a0.media_url AS "MediaUrl", a0.file_size AS "FileSize", a0.announcement_id
    FROM master.announcement_documents AS a0
    WHERE a0.is_active
) AS t0 ON t.id = t0.announcement_id
ORDER BY t.id
```

**Comment by roji** [comment]
Please post a runnable, minimal code sample that shows this happening.

**Comment by jxwaters** [comment]
@roji  - I can work on that - but this project uses the Specification pattern, and a lot of other things, so I can't easily extract this - I will need to create my own mini project with just this in it. I can do that if you can verify first that sorting included child collections is supported in the PostgreSQL driver?

**Comment by roji** [comment]
Here you go - see the code sample below, which shows a trivial case where the ordering appears in the SQL.

BTW It's almost never useful to open an issue without a repro... If the bug really occurs in trivial cases, then it should be very easy for the reporter to just attach a minimal repro. And if it doesn't, then there's nothing the maintainer can do anyway, since they need code which shows the error happening.

<details>
<summary>Attempted repro</summary>

```c#
await using var ctx = new BlogContext();
await ctx.Database.EnsureDeletedAsync();
await ctx.Database.EnsureCreatedAsync();

_ = ctx.Blogs
    .Include(b => b.Posts.OrderByDescending(p => p.Created))
    .ToList();

public class BlogContext : DbContext
{
    public DbSet<Blog> Blogs { get; set; }

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder
            .UseNpgsql(@"Host=localhost;Username=test;Password=test")
            .LogTo(Console.WriteLine, LogLevel.Information)
            .EnableSensitiveDataLogging();
}

public class Blog
{
    public int Id { get; set; }
    public required string Name { get; set; }

    public List<Post> Posts { get; set; }
}

public class Post
{
    public int Id { get; set; }
    public DateTime Created { get; set; }

    public Blog Blog { get; set; }
}
```

SQL:

```sql
SELECT b."Id", b."Name", p."Id", p."BlogId", p."Created"
FROM "Blogs" AS b
LEFT JOIN "Post" AS p ON b."Id" = p."BlogId"
ORDER BY b."Id", p."Created" DESC
```

</details>

**Comment by ajcvickers** [comment]
@jxwaters Make sure the Include is actually being used by the final query; check your logs for warnings.

**Comment by pinkfloydx33** [comment]
Are you using AutoMapper by any chance?

I ask because when using Automapper with ProjectTo, the includes will be ignored outright, with AM adding its own "includes" based on the projection dto. I'm calling this out as I had a similar case and it took me a while to realize AM was the culprit.

**Comment by jxwaters** [comment]
Yes, I am - and the Includes are not being ignored, just the OrderBy on them - i.e. I get the child collections when I add the includes, and I don not when I don't, as expected. But the orderby isn't honored.

**Comment by pinkfloydx33** [comment]
Let me rephrase that:

Assuming that you are using `ProjectTo`, remove the `Include` calls from your code and you should see that  `ProjectTo` automatically adds them back in. It does this based on your mapping profiles and how the `MapFrom` are configured for the relationships (either explicit mappings or implicit by property name). It is part of how automapper _rewrites_ the underlying expression tree _based on your mapping configurations_. It will automatically add "includes" for properties which _are_ part of the projection to the destination DTO (assuming they are in the mapping profile). As part of this it simply "discards" any explicit `Include` calls you make and replaces them with the ones it deduces from the target map. 

This is actually pretty much the **entire point** of using `ProjectTo`. 

If you want the `OrderBy`, you need to include it in your mapping profile.

```csharp 
CreateMap<Source, Dest>() // or CreateProjection
  .ForMember(
     dest => dest.CollectionProperty, 
     opt => opt.MapFrom(src => src.EntityCollection.OrderBy(c => c.Whatever)) 
  );
```

The only way to _prevent_ it from auto-including during a projection is to use the `ExplicitExpansion` options. If you need special order-by logic, you can _parameterize the map_ and pass parameters to `ProjectTo`. The Automapper docs have examples of both things.

I can assure you this is not an EF bug, but is rather a  _feature_ of AutoMapper's `ProjectTo`.

Your other option is to query using normal EF entity types (omitting the `ProjectTo`), materializing to a list and then doing a normal map in memory. 

If you are _not_ using `ProjectTo`, then I apologize and it certainly sounds like an EF issue.

**Comment by stevendarby** [comment]
Note that ProjectTo just builds a select expression. You would see the same with Select so it's not unique to AutoMapper / ProjectTo.

**Comment by pinkfloydx33** [comment]
Also true, except with AM it's not so obvious that that is what you are doing. I actually fell into this trap a while ago and once I realized what it's doing, it made a lot of sense. That time spent is why I recognized AM as the culprit off the bat. 

SFAIK Include only works if you are loading an entity, into your actual entity models. Once a select projection is involved (manually or via ProjectTo) you lose that since you aren't loading the EF types anymore as you're being explicit about the shape of the request. It's the same reason (I think) that the results of such a query wouldn't get tracked.

**Comment by roji** [comment]
@pinkfloydx33 it's possible to use Select to project out (tracked) EF entities.

In any case, beyond all the discussion, what we need here is a minimal repro which shows the issue happening.

**Comment by jxwaters** [comment]
Thanks for the details, @pinkfloydx33 . I get what you are saying. However - there are two additional data points to why I am doing it this way. First - I use ProjectTo because many times there are lots of fields in the Entities that I don't want in my DTOs, so I don't want them materialized in memory and then discarded - ProjectTo is perfect for making 'slim' SQL queries instead of fetching full entities and then mapping them and removing unused fields. In THIS case I also want a child collection, and these also have properties I am not interested in having in my DTO. One of these I want to sort by. I want to sort the child collection by Created (a date/time), but I don't want to include Created in my DTO. So I want this all to happen in the generated SQL. I will try your suggesting of putting this in the  mapper, although it seems to me that goes beyond mapping.

**Comment by jxwaters** [comment]
@pinkfloydx33 I can confirm you are correct - 

I removed
       .Include(x => x.Documents.OrderByDescending( d => d.Created));

From the Spec, and added

        CreateMap<Announcement, AnnouncementDetailDto>()
            .ForMember(
                dest => dest.Documents,
                opt => opt.MapFrom(
                    src => src.Documents.OrderByDescending(c => c.Created))
            );

To the mapping.... and now they are sorted correctly

**Comment by pinkfloydx33** [comment]
I found it a little unintuitive at first as well and I totally understand why you are using ProjectTo. AM considers projection and mapping two different things while at the same time using the same system to configure both of them. It's unfortunate, especially if you want to mix projections and nornal mapping, or if you want to order/filter your child collections (beware, filtering the collection suffers the same caveats and is when you'd likely need a parameterized map/projection) 

All in, it completely rewrites all the Select expressions that EF builds up and only cares about the mapping configuration, filters on the root expression and nothing more. I can't find it now, but at the time I recall reading an AM issue describing the same problem to which the maintainer responded "working as intended"

**Comment by roji** [comment]
@pinkfloydx33 @jxwaters at the end of the day, EF receives a LINQ expression tree as its query input; EF itself isn't aware of Automapper, which AFAIK simply transforms the query tree before that tree ends up in EF (note that I've never actually used Automapper, so my mental model could be wrong). So the thing to generally do with any Automapper difficulties is to isolate the final expression tree that Automapper produces, and to make sure it makes sense and corresponds to what you want. If you see an AM-produced tree which makes sense, but which EF somehow doesn't process in a way which you want, you can then provide a minimal repro, ideally with AM cut out - simply passing the tree directly to EF as-is.

I'd also recommend considering whether slimming down projected properties is a good enough reason to involve AM in a project; I have nothing against and and am not saying it's bad in any way, but it is an additional layer that introduces its complexities (at minimum, understanding exactly which tree gets produced for EF). Using a simple Select is usually satisfactory for reducing the set of properties, so that may be an easier way to achieve the same thing (of course, more complicated mapping/projection scenarios may indeed justify AM).

In any case, it seems from the above that there isn't anything wrong on the EF side - please confirm if this still needs to remain open.

**Comment by pinkfloydx33** [comment]
@roji  I realize that, and it's the point I was trying to make to the OP (in perhaps _way_ too many words 😃): AM's `ProjectTo` _intentionally_ alters the underlying expressions tree, discards the `Include` calls and creates its own Select projection. A comparison of the DebugView before/after calling `ProjectTo` would highlight that the new expression is considerably different from the original, thus the expected behavior is also different.

To quote myself:

> I can assure you this is not an EF bug, but is rather a _feature_ of AutoMapper's `ProjectTo`.

**Comment by stevendarby** [comment]
Seems like the issue is resolved but some further discussion may still be of interest to others reading...

> I'd also recommend considering whether slimming down projected properties is a good enough reason to involve AM in a project

> of course, more complicated mapping/projection scenarios may indeed justify AM

FWIW the opposite is often advised by Jimmy Bogard and co 😀 

AutoMapper is at its best when it's simply _automatically_ mapping properties of the same name (e.g. from an entity to a DTO with a subset of properties). The more manual mapping configuration is required, the more it's encouraged to avoid AM and write your own projection.

**Comment by roji** [comment]
OK - I'm indeed no expert on Automapper.

In any case, as there doesn't seem to be an EF issue here, I'm going to go ahead and close this. If you think something needs are attention, please post back here and we'll revisit.
